### PR TITLE
[T2I LoRA training] fix: unscale fp16 gradient problem

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Fixes: https://github.com/huggingface/diffusers/issues/6086

### CODE REVIEW

File: examples/text_to_image/train_text_to_image_lora.py [hunk 1]
```diff
@@ -919,39 +923,42 @@ def collate_fn(examples):
...
 
-    # Final inference
-    # Load previous pipeline
-    pipeline = DiffusionPipeline.from_pretrained(
-        args.pretrained_model_name_or_path, revision=args.revision, variant=args.variant, torch_dtype=weight_dtype
-    )
-    pipeline = pipeline.to(accelerator.device)
+        # Final inference
+        # Load previous pipeline
+        if args.validation_prompt is not None:
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:928]: If not `validation_prompt` was passed we must not run this step.

File: examples/text_to_image/train_text_to_image_lora.py [hunk 2]
```diff
@@ -919,39 +923,42 @@ def collate_fn(examples):
...
+                revision=args.revision,
+                variant=args.variant,
+                torch_dtype=weight_dtype,
+            )
+            pipeline = pipeline.to(accelerator.device)
 
-    # load attention processors
-    pipeline.unet.load_attn_procs(args.output_dir)
+            # load attention processors
+            pipeline.load_lora_weights(args.output_dir)
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:938]: Make sure to use `load_lora_weights()` instead of `load_attn_procs()`.

**Comment by sr5434** [comment]
@sayakpaul I am getting this error for regular LoRA fine-tune:
```
Steps:   0% 0/1500 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/content/diffusers/examples/text_to_image/train_text_to_image_lora_sdxl.py", line 967, in <module>
    main()
  File "/content/diffusers/examples/text_to_image/train_text_to_image_lora_sdxl.py", line 774, in main
    model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/utils/operations.py", line 680, in forward
    return model_forward(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/utils/operations.py", line 668, in __call__
    return convert_to_fp32(self.model_forward(*args, **kwargs))
  File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 16, in decorate_autocast
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/diffusers/models/unet_2d_condition.py", line 1004, in forward
    if "text_embeds" not in added_cond_kwargs:
TypeError: argument of type 'NoneType' is not iterable
```
This is in a free GPU enabled Google Colab

**Comment by sayakpaul** [comment]
I am not sure what script you're using here.

**Review by younesbelkada — APPROVED** [review]
Makes sense thanks! in the future we could also expose a method on PEFT to upcast trainable params in fp32 ! cc @BenjaminBossan @pacman100 similarly as `prepare_model_for_kbit_training`

**Comment by younesbelkada** [comment]
>  Is it only when the user explicitly loads a model in float16?

@sayakpaul can confirm but I think that's the case right ?

**Comment by sayakpaul** [comment]
> > Is it only when the user explicitly loads a model in float16?
> 
> @sayakpaul can confirm but I think that's the case right ?

Indeed that's the case. Only reduced precisions.

**Comment by patrickvonplaten** [comment]
@patil-suraj @williamberman can you please also take a look here?

**Comment by sr5434** [comment]
> I am not sure what script you're using here.

I am using the ```train_text_to_image_lora_sdxl.py``` script

**Comment by BenjaminBossan** [comment]
> Indeed that's the case. Only reduced precisions.

Does this also apply to bf16? If not, I think the dtype conversion should be conditional, i.e. `if args.mixed_precision == "fp16"`.

**Comment by sayakpaul** [comment]
@BenjaminBossan done in https://github.com/huggingface/diffusers/pull/6119/commits/8ac462ba8552c9cdd48b1111da616c8bb26c6583.

**Comment by patrickvonplaten** [comment]
Hmm, but now we're just silently disabling fp16 training - didn't this work before (e.g. that the whole UNet is kept in fp16 when LoRA is trained). Why doesn't it work anymore?

**Comment by patrickvonplaten** [comment]
The problem here is the following IMO:
- We move both LoRA weights and non-LoRA weights to fp16 before training; then in mixed precision training the LoRA weights should **not** be in FP16 and thus an error is thrown
- IMO, the solution should not be to move **all** weights (including non-trainable weights) to full fp32, instead we should only move the trainable LoRA weights to fp32 and keep the rest in fp16 to not blow up memory

**Comment by younesbelkada** [comment]
the changes proposed only upcasts the LoRA in fp32 (with the check `requires_grad`) and before the PEFT integration all LoRA layers were in fp32 because the arg `dtype` was never used: https://github.com/huggingface/diffusers/blob/a0c54828a170a03947dc6f7f44aa6548d288f36b/src/diffusers/models/lora.py#L204 
```python
    for attn_processor_name, attn_processor in unet.attn_processors.items():
        # Parse the attention module.
        attn_module = unet
        for n in attn_processor_name.split(".")[:-1]:
            attn_module = getattr(attn_module, n)

        # Set the `lora_layer` attribute of the attention-related matrices.
        attn_module.to_q.set_lora_layer(
            LoRALinearLayer(
                in_features=attn_module.to_q.in_features, out_features=attn_module.to_q.out_features, rank=args.rank
            )
        )
        attn_module.to_k.set_lora_layer(
            LoRALinearLayer(
                in_features=attn_module.to_k.in_features, out_features=attn_module.to_k.out_features, rank=args.rank
            )
        )
        attn_module.to_v.set_lora_layer(
            LoRALinearLayer(
                in_features=attn_module.to_v.in_features, out_features=attn_module.to_v.out_features, rank=args.rank
            )
        )
```

**Comment by younesbelkada** [comment]
one cleaner check could be to check if the module is an instance of `BaseTunerLayer` and upcast it if that's the case

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 3]
```diff
@@ -1024,6 +1024,16 @@ def main(args):
...
         text_encoder_two.add_adapter(text_lora_config)
 
+    # Make sure the trainable params are in float32.
+    if args.mixed_precision == "fp16":
+        models = [unet]
+        if args.train_text_encoder:
+            models.extend([text_encoder_one, text_encoder_two])
+        for model in models:
+            for param in model.parameters():
+                if param.requires_grad:
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:1034]:
```suggestion
                if param.requires_grad:
```
```suggestion
                # only upcast trainable parameters (LoRA) into fp32
                if param.requires_grad:
```

File: examples/text_to_image/train_text_to_image_lora.py [hunk 4]
```diff
@@ -493,7 +493,12 @@ def main():
     vae.to(accelerator.device, dtype=weight_dtype)
     text_encoder.to(accelerator.device, dtype=weight_dtype)
 
+    # Add adapter and make sure the trainable params are in float32.
     unet.add_adapter(unet_lora_config)
+    if args.mixed_precision == "fp16":
+        for param in unet.parameters():
+            if param.requires_grad:
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:500]:
```suggestion
            # only upcast trainable parameters (LoRA) into fp32
            if param.requires_grad:
```

**Comment by patrickvonplaten** [comment]
> the changes proposed only upcasts the LoRA in fp32 (with the check `requires_grad`. When you inject adapters the non-LoRA weights will have `requires_grad` set to `False`) and before the PEFT integration all LoRA layers were in fp32 because the arg `dtype` was never used in the example scripts:
> 
> https://github.com/huggingface/diffusers/blob/a0c54828a170a03947dc6f7f44aa6548d288f36b/src/diffusers/models/lora.py#L204
> 
> ```python
>     for attn_processor_name, attn_processor in unet.attn_processors.items():
>         # Parse the attention module.
>         attn_module = unet
>         for n in attn_processor_name.split(".")[:-1]:
>             attn_module = getattr(attn_module, n)
> 
>         # Set the `lora_layer` attribute of the attention-related matrices.
>         attn_module.to_q.set_lora_layer(
>             LoRALinearLayer(
>                 in_features=attn_module.to_q.in_features, out_features=attn_module.to_q.out_features, rank=args.rank
>             )
>         )
>         attn_module.to_k.set_lora_layer(
>             LoRALinearLayer(
>                 in_features=attn_module.to_k.in_features, out_features=attn_module.to_k.out_features, rank=args.rank
>             )
>         )
>         attn_module.to_v.set_lora_layer(
>             LoRALinearLayer(
>                 in_features=attn_module.to_v.in_features, out_features=attn_module.to_v.out_features, rank=args.rank
>             )
>         )
> ```

I see that makes sense! Thanks for the explanation

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sayakpaul** [comment]
I pulled in the changes from this PR and added to https://github.com/huggingface/diffusers/pull/6225.

```diff
         text_encoder_one.add_adapter(text_lora_config)
         text_encoder_two.add_adapter(text_lora_config)

+    # Make sure the trainable params are in float32.
+    if args.mixed_precision == "fp16":
+        models = [unet]
+        if args.train_text_encoder:
+            models.extend([text_encoder_one, text_encoder_two])
+        for model in models:
+            for param in model.parameters():
+                # only upcast trainable parameters (LoRA) into fp32
+                if param.requires_grad:
+                    param.data = param.to(torch.float32)
+
     # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
     def save_model_hook(models, weights, output_dir):
         if accelerator.is_main_process:
```

I can confirm that things are working well: https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/ow1vrez8. See the "test" media pictures. 

Command I ran:

```bash
CUDA_VISIBLE_DEVICES=1 accelerate launch train_with_fixes.py \
  --pretrained_model_name_or_path="stabilityai/stable-diffusion-xl-base-1.0" \
  --pretrained_vae_model_name_or_path="madebyollin/sdxl-vae-fp16-fix" \
  --instance_data_dir="dog" \
  --output_dir="corgy_dog_LoRA" \
  --mixed_precision="fp16" \
  --instance_prompt="a photo of TOK dog" \
  --resolution=1024 \
  --train_batch_size=4 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --snr_gamma=5.0 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --mixed_precision="fp16" \
  --use_8bit_adam \
  --max_train_steps=500 \
  --checkpointing_steps=100 \
  --push_to_hub \
  --validation_prompt="a photo of TOK dog in a bucket at the beach" \
  --report_to="wandb" \
  --seed="0"
```

Trained model: https://huggingface.co/sayakpaul/corgy_dog_LoRA. I am gonna try to run using Colab free tier too and report back here.

**Comment by sayakpaul** [comment]
To the ones wondering if this stuff would run on free-tier Colab Notebook, https://colab.research.google.com/gist/sayakpaul/9615b89369f3ef23cc29d0dac58253dd/scratchpad.ipynb should clear all the doubts once and for all 💪

**Comment by yashveer08** [comment]
> To the ones wondering if this stuff would run on free-tier Colab Notebook, https://colab.research.google.com/gist/sayakpaul/9615b89369f3ef23cc29d0dac58253dd/scratchpad.ipynb should clear all the doubts once and for all 💪

This seems to be working, but when added the ```metadata.jsonl``` file, the dataset library is causing the issue.
It shows the below error.
can you cross check your colab with the caption file @sayakpaul , it would be great help.
Screenshot 2023-12-19 at 12 13 26 PM

**Comment by sayakpaul** [comment]
That is an unrelated problem and you should instead file this in the `datasets` repo.

**Comment by yashveer08** [comment]
> That is an unrelated problem and you should instead file this in the `datasets` repo.

Sure will do, but is there a way apart from this to train an SDXL model with captions for each image, alternative to this?
@sayakpaul

**Comment by mattfreed** [comment]
I am also receiving an error after applying this fix. Gives me a SIGKILL: 9. --- Error is not due to any compute or storage constraints. Training on colab with a T4. The safe tensor DO save ( have yet to try it on inference). I also noticed when I did not have the Datasets library installed (using a custom local dataset), the "loading pipeline components" would only get to 57% vs with it installed getting to 86%. 

```
Loading pipeline components...:  86%|████████▌ | 6/7 [00:17<00:01,  1.87s/it]{'attention_type', 'dropout', 'reverse_transformer_layers_per_block'} was not found in config. Values will be initialized to default values.


Loading pipeline components...:  86%|████████▌ | 6/7 [00:28<00:01,  1.87s/it]Traceback (most recent call last):
  File "/usr/local/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/accelerate_cli.py", line 47, in main
    args.func(args)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/launch.py", line 1017, in launch_command
    simple_launcher(args)
  File "/usr/local/lib/python3.10/dist-packages/accelerate/commands/launch.py", line 637, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'train_dreambooth_lora_sdxl.py', '--pretrained_model_name_or_path=stabilityai/stable-diffusion-xl-base-1.0', '--pretrained_vae_model_name_or_path=madebyollin/sdxl-vae-fp16-fix', '--instance_data_dir=./user_images', '--output_dir=./weights', '--caption_column=prompt', '--mixed_precision=fp16', '--instance_prompt=a photo of a [username] person', '--resolution=512', '--train_batch_size=6', '--gradient_accumulation_steps=3', '--gradient_checkpointing', '--learning_rate=1e-4', '--snr_gamma=5.0', '--lr_scheduler=constant', '--lr_warmup_steps=0', '--mixed_precision=fp16', '--use_8bit_adam', '--max_train_steps=1', '--checkpointing_steps=717', '--seed=0']' died with <Signals.SIGKILL: 9>.

Training failed, weights not saved for [username]
```

This is the setup I am using:
```
`def launch_training_script(username, output_dir):
    command = [
            "accelerate", "launch", "train_dreambooth_lora_sdxl.py",
            "--pretrained_model_name_or_path=stabilityai/stable-diffusion-xl-base-1.0",
            "--pretrained_vae_model_name_or_path=madebyollin/sdxl-vae-fp16-fix",
            "--instance_data_dir=./user_images",
            f"--output_dir={output_dir}",
            "--caption_column=prompt",
            "--mixed_precision=fp16",
            f"--instance_prompt=a photo of a {username} person",
            "--resolution=512",
            "--train_batch_size=6",
            "--gradient_accumulation_steps=3",
            "--gradient_checkpointing",
            "--learning_rate=1e-4",
            "--snr_gamma=5.0",
            "--lr_scheduler=constant",
            "--lr_warmup_steps=0",
            "--mixed_precision=fp16",
            "--use_8bit_adam",
            "--max_train_steps=1",
            "--checkpointing_steps=100",
            "--seed=0"
        ]

    # Launch the script and wait for it to complete
    result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)

    # Output the result
    if result.returncode == 0:
        print("Script completed successfully")
        print(result.stdout)
        return True  # Indicate success
    else:
        print("Script failed with errors")
        print(result.stderr)
        return False  # Indicate failure`
```

**Comment by sayakpaul** [comment]
This you will have to debug your way, cause it's not exactly the same code that you're using.

**Pull request merged**